#!/bin/bash

# to extract the vcf of neutral regions I create beds of all the regions I do
# not consider neutral, such as:
# 1) all genes
# 2) all outlier windows from baypass
# 3) all outlier snps from rda

vcf=1-Preparing_Genetic_Data/tables/ll_wholegenome_LyCa_ref.sorted.filter7.vcf
vcf_neutral=4-Downstream_Analyses/tables/ll_wholegenome_LyCa_ref.sorted.filter7.${1}.neutral.vcf

## 1. all genes 
# bed generated by Dani as follows:
# cd /GRUPOS/grupolince/reference_genomes/lynx_canadensis
# 
# awk -F"\t" '$3 == "gene" {printf ("%s\t%s\t%s\n", $1, $4-5001, $5+5000)}' lc4.NCBI.nr_main.gff3 \
#  > lc4.NCBI.nr_main.genes.plus5000.temp_bed
# 
# join -1 1 -2 1 <(LANG=en_EN sort -k1,1 -k2,2n -k3,3n lc_ref_all_the_genome.bed) \
#  <(LANG=en_EN sort -k1,1 -k2,2n -k3,3n lc4.NCBI.nr_main.genes.plus5000.temp_bed) | 
#  awk -v OFS='\t' '{if ($4<0) {$4="1"} else {$4=$4}; print}' | 
#  awk -v OFS='\t' '{if ($5>$3) {$5=$3} else {$5=$5}; print $1,$4,$5}' | 
#  bedtools merge -i stdin -d 1 > lc4.NCBI.nr_main.genes.plus5000.bed
# 
# rm lc4.NCBI.nr_main.genes.plus5000.temp_bed

# all genes bed:
genes_bed=/GRUPOS/grupolince/reference_genomes/lynx_canadensis/lc4.NCBI.nr_main.genes.plus5000.bed

# remove genes
echo "removing genes"

cd /home/ebazzicalupo/Local_adaptation_Eurasian_lynx

bedtools subtract -header \
 -a $vcf \
 -b $genes_bed \
 > $vcf_neutral

# then:
# 2. all outlier windows from baypass
# should be axessed from the list of included variables:
var_list=($(echo "${1}" | cut -d'.' -f1 | cut -d'_' -f2- | tr '+' ' '))
echo "removing univariate results of variables:"
echo "${var_list[@]}"

# so remove them one by one
for var in ${var_list[@]}
 do

  echo "$var"
  
  baypass_bed=3-Identifying_Candidate_Loci/tables/${var}_genwin_windows_outliers.bed

  bedtools subtract -header \
   -a $vcf_neutral \
   -b $baypass_bed \
   > tmp && mv tmp $vcf_neutral

done

# then:
# 3. all outlier snps from rda
if [[ $(echo "${1}" | cut -d'.' -f2) == "cond_PC1+PC2" ]]
  then
  K=2
  elif [[ $(echo "${1}" | cut -d'.' -f2) == "cond_PC2" ]]
  then
  K=3
fi
echo "removing candidates from multivariate table:"
echo "${1}.cand_sd.K${K}.candidates.tsv"

# are found in table:
rda_table=3-Identifying_Candidate_Loci/tables/${1}.cand_sd.K${K}.candidates.tsv
# that can be converted to bed
rda_bed=3-Identifying_Candidate_Loci/tables/${1}.cand_sd.K${K}.candidates.bed
cat $rda_table | grep -v "snp" | cut -f2 | sed 's/"//g' | tr ':' '\t' | 
 rev |  cut -d'_' -f2- | rev |
 awk -v OFS='\t' '{print $1, $2-1, $2}' > $rda_bed
# and then subtracted from neutral vcf
bedtools subtract -header \
   -a $vcf_neutral \
   -b $rda_bed \
   > tmp && mv tmp $vcf_neutral
